lua/luasnip/util/environ.lua

```diff
@@ -1,190 +1,180 @@
-	return vim.fn.expand("%:t:s?\\.[^\\.]\\+$??")
-end
+local Environ = {}
 
-function lazy_vars.TM_DIRECTORY()
-	return vim.fn.expand("%:p:h")
-end
-function lazy_vars.TM_FILEPATH()
-	return vim.fn.expand("%:p")
+-- returns nil, but that should be alright.
+-- If not, use metatable.
+function Environ.is_table(key)
+	return builtin_vars.is_table[key]
 end
 
--- Vscode only
+function Environ:new(pos, o)
+	o = o or {}
+	setmetatable(o, self)
 
-function lazy_vars.CLIPBOARD() -- The contents of your clipboard
-	return vim.fn.getreg('"', 1, true)
-end
+	for ns_name, ns in pairs(namespaces) do
+		local eager_vars = {}
+		if ns.init then
+			eager_vars = ns.init(pos)
+		end
+		for _, eager in ipairs(ns.eager) do
+			if not eager_vars[eager] then
+				eager_vars[eager] = ns.vars(eager)
+			end
+		end
 
-local function buf_to_ws_part()
-	local LSP_WORSKPACE_PARTS = "LSP_WORSKPACE_PARTS"
-	local ok, ws_parts = pcall(vim.api.nvim_buf_get_var, 0, LSP_WORSKPACE_PARTS)
-	if not ok then
-		local file_path = vim.fn.expand("%:p")
+		local prefix = ""
+		if ns_name ~= "" then
+			prefix = ns_name .. "_"
+		end
 
-		for _, ws in pairs(vim.lsp.buf.list_workspace_folders()) do
-			if file_path:find(ws, 1, true) == 1 then
-				ws_parts = { ws, file_path:sub(#ws + 2, -1) }
-				break
+		for name, val in pairs(eager_vars) do
+			name = prefix .. name
+			local val_type = type(val)
+			if val and val_type ~= "string" and val_type ~= "table" then
+				val = tostring(val)
 			end
+			rawset(o, name, val)
 		end
-		-- If it can't be extracted from lsp, then we use the file path
-		if not ok and not ws_parts then
-			ws_parts = { vim.fn.expand("%:p:h"), vim.fn.expand("%:p:t") }
-		end
-		vim.api.nvim_buf_set_var(0, LSP_WORSKPACE_PARTS, ws_parts)
 	end
-	return ws_parts
-end
-
-function lazy_vars.RELATIVE_FILEPATH() -- The relative (to the opened workspace or folder) file path of the current document
-	return buf_to_ws_part()[2]
-end
-
-function lazy_vars.WORKSPACE_FOLDER() -- The path of the opened workspace or folder
-	return buf_to_ws_part()[1]
+	return o
 end
 
-function lazy_vars.WORKSPACE_NAME() -- The name of the opened workspace or folder
-	local parts = vim.split(buf_to_ws_part()[1] or "", "[\\/]")
-	return parts[#parts]
-end
+local builtin_ns_names = vim.inspect(vim.tbl_keys(builtin_vars.builtin_ns))
+
+function Environ.env_namespace(name, namespace)
+	assert(
+		#name > 0 and not (name:find("_")),
+		("You can't create a namespace with name '%s' empty nor containing _"):format(
+			name
+		)
+	)
+	assert(
+		not builtin_vars.builtin_ns[name],
+		("You can't create a namespace with name '%s' because is one one of %s"):format(
+			name,
+			builtin_ns_names
+		)
+	)
+	local ns = namespace
+
+	assert(
+		ns and type(ns) == "table",
+		("Your namespace '%s' has to be a table"):format(name)
+	)
+	assert(
+		ns.init or ns.vars,
+		("Your namespace '%s' needs init or vars"):format(name)
+	)
+	assert(
+		not namespace.eager and ns.vars,
```

Comment:
I think this is a bit hard to decipher (at least for me), what do you think of a (well, slightyl) more explicit `namespace.eager and ns.vars or true` (I wish, for the first time, that "implies" was implemented in more (any?) languages)
(or maybe just add a comment that `not namespace.eager and ns.vars` is `implies`)

Reply:
> 

That's the same I said 😂

Reply:
Yeah pretty much I know xD
Just wanted to list it all again so it's clear but still show the difference with `nil`

Reply:
(The difference being "no default", but now I doubt it's the best design, a default might be better)